(SOL-1154) Enabling search in devstack

marjev opened the pull request:
JIRA ticket: https://openedx.atlassian.net/browse/SOL-1154

- [x] 1st review: @dino-cikatic 
- [x] 2nd review: @sarina

edx-webhook commented:
Thanks for the pull request, @marjev! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request. 

To automatically create an OSPR issue for this pull request, just visit this link: https://openedx-webhooks.herokuapp.com/github/process_pr?repo=edx%2Fedx-platform&number=9478

edx-webhook commented:
Thanks for the pull request, @marjev! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request. 

To automatically create an OSPR issue for this pull request, just visit this link: https://openedx-webhooks.herokuapp.com/github/process_pr?repo=edx%2Fedx-platform&number=9478

marjev commented:
jenkins run python

vkaracic commented:
jenkins run js
jenkins run python

marjev commented:
jenkins run bokchoy

marjev commented:
@dino-cikatic please review this

@sarina could you be a 2nd reviewer for this?
Related to JIRA story: https://openedx.atlassian.net/browse/SOL-1154

This fix should address the problem of courses not being searchable on devstack out of the box.

Fix can be tested on:
1) fresh devstack environment, where courses that come with devstack by default (currently only edX Demonstation Course) will be reindexed when ElasticSearch index is available and empty upon running LMS or Studio using `paver devstack` command. After the course has been indexed, ElasticSearch index mapping should have body, which basically means there's content available for searching, thus preventing the further indexing upon powering up LMS or studio.

2) devstack environment that has courses in modulestore, but none of the courseware content indexed in the ElasticSearch. Applying this fix will index content of all existing courses, which can take a while depending on number of courses in the modulestore and their content size.

3) devstack environment, that has some courseware content indexed, by deleting index using `curl -X DELETE http://localhost:9200/courseware_index` command as edxapp user. After running either `paver devstack lms` or `paver devstack studio`, you should see courses being indexed, since nothing is indexed.

dino-cikatic commented on cms/djangoapps/contentstore/management/commands/reindex_course.py:
```diff
@@ -0,0 +1,111 @@
+        try:
+            result = CourseKey.from_string(raw_value)
+        except InvalidKeyError:
+            result = SlashSeparatedCourseKey.from_deprecated_string(raw_value)
```
This block can except two times, throwing exception again in except block. I think this was properly addressed in delete_course command. 
Also, since this pattern is used multiple times across platform (especially commands), we might remove duplicated code and simplify course_key parsing changes in future.

sarina replied:
This pattern is no longer the correct pattern to use. You should just use `course_key = CourseKey.from_string(course_id)`

See https://github.com/edx/edx-platform/wiki/Opaque-Keys-(Locators)#basics

dino-cikatic commented on cms/djangoapps/contentstore/management/commands/tests/test_reindex_courses.py:
```diff
@@ -0,0 +1,130 @@
+            org="test", library="lib2", display_name="run2", default_store=ModuleStoreEnum.Type.split
+        )
+
+        self.first_course = CourseFactory.create(
```
Is there a reason we insist on a split store here? For Library it has to be split, but I don't think we need to hardcode this for courses.

sarina commented on cms/djangoapps/contentstore/management/commands/reindex_course.py:
```diff
@@ -0,0 +1,111 @@
+        except InvalidKeyError:
+            result = SlashSeparatedCourseKey.from_deprecated_string(raw_value)
+
+        if not isinstance(result, CourseLocator):
```
I don't think this will be necessary with the correct pattern - https://github.com/edx/edx-platform/wiki/Opaque-Keys-(Locators)#basics

marjev replied:
I'm covering the case when someone passes a library key instead of a course key as a parameter when invoking this command. This will prevent indexing errors.

sarina commented on pavelib/paver_tests/test_servers.py:
```diff
@@ -83,13 +86,27 @@ def test_devstack(self, server_options):
         options["system"] = "lms"
+        options["expected_messages"] = [
+            EXPECTED_INDEX_COURSE_COMMAND.format(
+                system="cms",
```
Why is this "cms" after options["system"] is set to "lms" in L93?

sarina commented on pavelib/paver_tests/test_servers.py:
```diff
@@ -83,13 +86,27 @@ def test_devstack(self, server_options):
+                system="cms",
+                settings=expected_settings,
+            )
+        ]
```
Why is this added to CMS? I thought the indexing was only for LMS

sarina commented on pavelib/servers.py:
```diff
@@ -135,6 +135,7 @@ def devstack(args):
     if args.optimized:
         settings = OPTIMIZED_SETTINGS
         asset_settings = OPTIMIZED_ASSETS_SETTINGS
+    sh(django_cmd('cms', settings, 'reindex_course', '--setup'))
```
Is this the right place to put this? Why are you indexing on the cms and not the lms? I'm confused why the commands live in the CMS but the settings are in lms/envs/devstack.py.

marjev replied:
All the logic regarding ElasticSearch indexing resides in the CMS, since course staff members will be indexing courses from there. You can see that in cms/envs/devstack.py we have a flag `ENABLE_COURSEWARE_INDEX`.

In lms/envs/devstack.py are the settings for enabling searching. That's the reason why the command lives in the CMS.

sarina commented:
I can't run this locally:
```
python manage.py cms --settings=devstack reindex_course --setup
Traceback (most recent call last):
  File "manage.py", line 116, in <module>
    execute_from_command_line([sys.argv[0]] + django_args)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 443, in execute_from_command_line
    utility.execute()
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 382, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/core/management/base.py", line 196, in run_from_argv
    self.execute(*args, **options.__dict__)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/core/management/base.py", line 232, in execute
    output = self.handle(*args, **options)
  File "/edx/app/edxapp/edx-platform/cms/djangoapps/contentstore/management/commands/reindex_course.py", line 77, in handle
    store = modulestore()
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/modulestore/django.py", line 206, in modulestore
    contentstore(),
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/contentstore/django.py", line 28, in contentstore
    _CONTENTSTORE[name] = class_(**options)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/contentstore/mongo.py", line 39, in __init__
    **kwargs
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/pymongo/mongo_client.py", line 369, in __init__
    raise ConnectionFailure(str(e))
pymongo.errors.ConnectionFailure: [Errno 111] Connection refused
```

Full traceback: https://gist.githubusercontent.com/sarina/373b8c752b3b7fa0a7b6/raw/1f184691b03083b570cb2943ac1f206995685190/gistfile1.txt

I can test if this works from an open source perspective but I am unfamiliar with Elasticsearch so I'm not certain I can fully be the best 2nd reviewer on this PR.

marjev commented:
@sarina Thanks for reviewing this! The error you got looks similar to the error that is raised when the vagrant instance wasn't halted properly. Can you confirm that removing mongo lock solves the issue?

I can tag someone from the solutions team to do the formal 2nd review from ElasticSearch perspective, but I would like you to be included, just so that we have someone looking at this from another (open source) perspective.

marjev commented:
@dino-cikatic I addressed the concerns you pointed out in your review, could you take a 2nd look and confirm that everything is done accordingly from ElasticSearch point of view?

dino-cikatic commented:
@marjev I know this is not in scope for this ticket, but changing these course key parses blocks in all commands might be worth doing. Can you maybe create a ticket to mark this work in future?
:+1: for PR.

sarina commented:
OK, this is working for me. Is there a way to configure not displaying the Course Type, for instances that don't use Course Type? Also, how are you determining Language of course?

![screen shot 2015-09-02 at 8 32 47 am](https://cloud.githubusercontent.com/assets/1985317/9631197/50ebf70a-514d-11e5-930a-6a61f2a64c77.png)

marjev commented:
@sarina I don't know if this will answer your question about Course Types, but if a course has a type, it will be indexed and it will be displayed. If an instance doesn't use Course Type (I'm assuming you mean that courses don't have a type), course type will not be displayed.

Course language is specified in the Schedule & Details Settings section in studio:

<img width="449" alt="screen shot 2015-09-02 at 15 14 17" src="https://cloud.githubusercontent.com/assets/6833568/9632117/87b7f742-5185-11e5-8a5a-14ab7491b02f.png">

The language is indexed with the course content.

sarina commented:
@marjev regarding the Course Type, all courses are I think by default "Honor". What I"m saying is if an installation doesn't use any course modes, "Honor" doesn't really make sense to show in the results and I think they would want to turn it off.

Regarding language, that does answer my question.

marjev commented:
@sarina you can override the default configuration by defining a new setting named COURSE_DISCOVERY_FILTERS which should be a list of facets you want to filter the search results by.

sarina commented:
Could your PR put in an example of that somewhere? or an explanation? I'm
not sure how an open source person would find out how to configure that.

On Wed, Sep 2, 2015 at 1:28 PM, Marko Jevtić <notifications@github.com>
wrote:

> @sarina <https://github.com/sarina> you can override the default
> configuration by defining a new setting named COURSE_DISCOVERY_FILTERS
> which should be a list of facets you want to filter the search results by.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/edx/edx-platform/pull/9478#issuecomment-137179740>.
>

marjev commented:
Yes, no problem. I'll figure something out.

edx-webhook commented:
Thanks for the pull request, @marjev! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

To automatically create an OSPR issue for this pull request, just visit this link: https://openedx-webhooks-test.herokuapp.com/github/process_pr?repo=edx%2Fedx-platform&number=9478

marjev commented:
@sarina, in commit a2963d80f3e5c6a03c3f5b0d6423dea2b3660d70, I added couple of comments, but I'm not certain whether they're descriptive enough.

I found documentation that describes the process of configuring search feature. It's available [here](https://github.com/edx/edx-documentation/blob/master/en_us/install_operations/source/configuration/edx_search.rst). I've opened a PR https://github.com/edx/edx-documentation/pull/526 that proposes an update on how to configure which facets will be displayed.

Document is available to open source community members [here](http://edx.readthedocs.org/projects/edx-installing-configuring-and-running/en/latest/configuration/edx_search.html).

sarina commented on lms/envs/common.py:
```diff
@@ -395,6 +395,9 @@
     'ENABLE_COURSE_DISCOVERY': False,
 
+    # Setting for overriding default filtering facets in Course discovery
+    # COURSE_DISCOVERY_FILTERS = ["org", "language", "modes"]
```
Why is this commented?

marjev replied:
I just placed that there as an example on how to override COURSE_DISCOVERY_FILTERS. By default, `["org", "language", "modes"]` is set on the edx-search side.

Someone can uncomment this and remove one or more filter facets to disable some of them, for example "modes" in the case you mentioned yesterday.

sarina commented on lms/envs/devstack.py:
```diff
@@ -167,7 +167,9 @@ def should_show_debug_toolbar(_):
 
-FEATURES['ENABLE_COURSE_DISCOVERY'] = False
+FEATURES['ENABLE_COURSE_DISCOVERY'] = True
+# Setting for overriding default filtering facets
```
I like the comment from common.py better: 

# Setting for overriding default filtering facets for Course discovery

marjev replied:
ah, sorry, I didn't even realize they are different, thanks! I'll change that

sarina commented:
@marjev Appreciate your work here. :+1: from me

marjev commented:
@dino-cikatic in commit 7a899a2e67c2b75dfc80276d405c9d51eaf33dc0, I addressed the issue with ElasticSearch, so please review this once more from ElasticSearch point of view.

dino-cikatic commented:
@marjev looks good.

marjev commented:
jenkins run a11y
jenkins run bokchoy

marjev commented:
@griffresch to verify this, please see the following comment: https://github.com/edx/edx-platform/pull/9478#issuecomment-136726822

edx-webhook commented:
Thanks for the pull request, @marjev! I've created [OSPR-2463](https://openedx.atlassian.net/browse/OSPR-2463) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).



If you like, you can add yourself to the [AUTHORS](https://github.com/edx/edx-platform/blob/master/AUTHORS)      file for this repo, though that isn't required.     Please see the [CONTRIBUTING](https://github.com/edx/edx-platform/blob/master/CONTRIBUTING.rst)     file for more information.
